### PR TITLE
Use Trajectory Planner For Move Primitives

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
Use the trajectory planner in the move primitives. This behaves effectively the same as the old, simple trajectory planner, but opens the door for passing polynomials to track to the move primitives from the navigator.

### Testing Done
Unit tests, simulated tests.

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [ ] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [ ] **Remove all commented out code**
- [ ] **Remove extra print statements**: for example, those just used for testing
- [ ] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [ ] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

**Review by jonathanlew — COMMENTED** [review]
Initial review: Can we pass the spline from the navigator to trajectory planner on the robots and follow that?

File: src/software/ai/hl/stp/play/corner_kick_play.cpp
Comment (jonathanlew) [inline src/software/ai/hl/stp/play/corner_kick_play.cpp]: why is this necessary?

File: src/firmware/app/primitives/move_helper.c
Comment (jonathanlew) [inline src/firmware/app/primitives/move_helper.c:32]: should rename movehelper to something related to trajectory planner

**Review by jonathanlew — CHANGES_REQUESTED** [review]
corner_kick_play_test works for me. Happy to see this coming together. There's a bit of clean up required in simulated_tests.

File: src/firmware/app/world/firmware_robot.h
Comment (MathewMacDougall) [inline src/firmware/app/world/firmware_robot.h]:
```suggestion
 * Get the absolute linear speed of this robot
```

File: src/software/simulated_tests/corner_kick_play_test.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/corner_kick_play_test.cpp]: this should be removed for the final submission (CI won't pass)

File: src/software/simulated_tests/example_play_test.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/example_play_test.cpp]: this should be removed for the final submission (CI won't pass)

File: src/software/simulated_tests/example_play_test.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/example_play_test.cpp]: Why did this change?

File: src/software/simulated_tests/simulated_test_fixture.cpp
Comment (MathewMacDougall) [inline src/software/simulated_tests/simulated_test_fixture.cpp]: commented out code

File: src/software/simulation/simulator.h
Comment (MathewMacDougall) [inline src/software/simulation/simulator.h]: comment without a function?

File: src/firmware/app/primitives/move_helper.c
Comment (MathewMacDougall) [inline src/firmware/app/primitives/move_helper.c:45]: Function seems unused? It also seems to be planning everything, not just the rotation?

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Nice. Not as big of a change as I was expecting

**Review by akhilveeraghanta — COMMENTED** [review]
Nothing else to add beyond existing comments
Exciting to see this coming together

File: src/software/ai/hl/stp/play/corner_kick_play.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/play/corner_kick_play.cpp]: Thanks for pointing this out, I think this was just a leftover from me messing around with stuff. It's not required, so removed.

File: src/software/simulated_tests/corner_kick_play_test.cpp
Comment (garethellis0) [reply src/software/simulated_tests/corner_kick_play_test.cpp]: removed

File: src/software/simulated_tests/example_play_test.cpp
Comment (garethellis0) [reply src/software/simulated_tests/example_play_test.cpp]: removed

File: src/software/simulated_tests/example_play_test.cpp
Comment (garethellis0) [reply src/software/simulated_tests/example_play_test.cpp]: whoops, just changed this so that the test would block and not end. Reverted.

File: src/firmware/app/primitives/move_helper.c
Comment (garethellis0) [reply src/firmware/app/primitives/move_helper.c:32]: Why do you say that? As I understand it this represents the state for the "Move Helper"?

File: src/firmware/app/primitives/move_helper.c
Comment (garethellis0) [reply src/firmware/app/primitives/move_helper.c:45]: The folds on github confused this man, this comment is not relevant :joy:

File: src/firmware/app/world/firmware_robot.h
Comment (garethellis0) [reply src/firmware/app/world/firmware_robot.h]: fixed!

File: src/software/ai/hl/stp/play/corner_kick_play.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/play/corner_kick_play.cpp]: Whoops, it's not, just a leftover from testing. Removed and looks like the simulated test is still fine.

File: src/software/simulated_tests/corner_kick_play_test.cpp
Comment (garethellis0) [reply src/software/simulated_tests/corner_kick_play_test.cpp]: removed

File: src/software/simulated_tests/example_play_test.cpp
Comment (garethellis0) [reply src/software/simulated_tests/example_play_test.cpp]: removed

File: src/software/simulated_tests/example_play_test.cpp
Comment (garethellis0) [reply src/software/simulated_tests/example_play_test.cpp]: Ah, just changed this so the test would block and not exit. Reverted!

File: src/software/simulated_tests/simulated_test_fixture.cpp
Comment (garethellis0) [reply src/software/simulated_tests/simulated_test_fixture.cpp]: removed

File: src/software/simulation/simulator.h
Comment (garethellis0) [reply src/software/simulation/simulator.h]: removed

**Comment by garethellis0** [comment]
> Initial review: Can we pass the spline from the navigator to trajectory planner on the robots and follow that?

Hmm, given that this is functional and doesn't reduce any regressions in it's current state, I think it might be a good idea to get this merged and do that in a future PR?

@jonathanlew

**Review by MathewMacDougall — DISMISSED** [review]
Nothing to add beyond remaining comments

File: src/firmware/app/primitives/move_helper.c
Comment (jonathanlew) [reply src/firmware/app/primitives/move_helper.c:32]: I meant that **`MoveHelper`** (not MoveHelperState_t) isn't a good name given that it uses the trajectory planner, so we can be more specific here

Comment (EvanMorcom) [reply src/firmware/app/primitives/move_helper.c:32]: I agree with both y'all. I don't know what we would really name it though. It's just a further abstraction around the TP data

File: src/firmware/app/primitives/move_helper.c
Comment (EvanMorcom) [inline src/firmware/app/primitives/move_helper.c]: This looks pretty garbled by clash format. Maybe we should turn it off and manually format

**Review by EvanMorcom — DISMISSED** [review]

**Comment by garethellis0** [comment]
@garethellis0 - Please finish this :fire: :cry:

**Comment by jonathanlew** [comment]
@garethellis0 if you fix merge conflicts, this should be good to go!

File: src/firmware/app/world/firmware_world.c
Comment (jonathanlew) [inline src/firmware/app/world/firmware_world.c:42]: The language here seems to imply that it's some sort of timestamp related to when the world was received, but it's actually the rtos time. We could specify rtoscurrenttime, but I don't see what this function has to do with the world

File: src/firmware/boards/legacy_robot_stm32f4/main.c
Comment (jonathanlew) [inline src/firmware/boards/legacy_robot_stm32f4/main.c:539]: interesting to use rtos time

**Review by jonathanlew — COMMENTED** [review]
some nits and questions about current time, otherwise looks pretty clean

File: src/firmware_new/boards/frankie_v1/freertos.c
Comment (garethellis0) [reply src/firmware_new/boards/frankie_v1/freertos.c]: whoops, spellcheck got aggressive in ubuntu 20.

File: src/software/ai/hl/stp/action/intercept_ball_action_test.cpp
Comment (garethellis0) [reply src/software/ai/hl/stp/action/intercept_ball_action_test.cpp:72]: It's a compiler error in the newer gcc version (hence why I caught it) https://stackoverflow.com/questions/2522299/c-catch-blocks-catch-exception-by-value-or-reference

File: src/firmware_new/boards/frankie_v1/freertos.c
Comment (garethellis0) [reply src/firmware_new/boards/frankie_v1/freertos.c]: done

File: src/firmware_new/boards/frankie_v1/main.c
Comment (garethellis0) [reply src/firmware_new/boards/frankie_v1/main.c]: done

File: src/software/backend/radio/mrf/messages.h
Comment (garethellis0) [reply src/software/backend/radio/mrf/messages.h]: done

File: src/software/simulated_tests/example_play_test.cpp
Comment (garethellis0) [reply src/software/simulated_tests/example_play_test.cpp]: whoops, it wasn't, reverted

File: src/firmware/app/world/firmware_world.c
Comment (garethellis0) [reply src/firmware/app/world/firmware_world.c:42]: Hmm. So `FirmwareWorld` is intended to be the abstraction for the app layer to interact with the rest of the world. It should not matter _how_ we got the time (it could be RTOS time, could be something entirely different, for example in sim). What are your thoughts on the way to best rephrase the language to make this clear?

File: src/firmware/boards/legacy_robot_stm32f4/main.c
Comment (garethellis0) [reply src/firmware/boards/legacy_robot_stm32f4/main.c:539]: It was the easiest way to get time in the firmware :man_shrugging:. A more complete design might be adding in a new abstraction for setting time before the app layer is ticked (so that a single tick of the app layer has the same time throughout), but I was really trying to avoid changes that couldn't be tested without a robot.

File: src/firmware/app/world/firmware_world.c
Comment (jonathanlew) [reply src/firmware/app/world/firmware_world.c:42]: ok would a time since power on abstraction work in sim?

File: src/software/ai/navigator/path_planner/path_planner_test.cpp
Comment (jonathanlew) [inline src/software/ai/navigator/path_planner/path_planner_test.cpp]: Can you explain the changes to these tests? They are failing in CI

**Review by jonathanlew — CHANGES_REQUESTED** [review]

File: src/software/ai/navigator/path_planner/path_planner_test.cpp
Comment (jonathanlew) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: they might have been caused by #1661

Comment (garethellis0) [reply src/software/ai/navigator/path_planner/path_planner_test.cpp]: Yep sorry. These break on gcc 9.3 (the default ubuntu 20 compiler) because the call to brace enclosed initializer lists is ambiguous. 

I've reverted the changes since they should pass CI, even if they don't pass on my machine.

**Review by jonathanlew — DISMISSED** [review]
LGTM pending CI!

**Comment by garethellis0** [comment]
Looks like we need another review here. @akhilveeraghanta ?

**Review by akhilveeraghanta — APPROVED** [review]
LGTM :fire:

**Review by jonathanlew — APPROVED** [review]

**Pull request merged**
